src/olympia/devhub/forms.py

```diff
@@ -568,6 +557,31 @@ def clean(self):
         return self.cleaned_data
 
 
+class SourceForm(WithSourceMixin, forms.ModelForm):
+    source = forms.FileField(required=False, widget=SourceFileInput)
+    has_source = forms.ChoiceField(
+        choices=(('yes', _('Yes')), ('no', _('No'))), required=True,
+        widget=RadioSelect)
+
+    class Meta:
+        model = Version
+        fields = ('source',)
+
+    def __init__(self, *args, **kwargs):
+        self.request = kwargs.pop('request')
+        super(SourceForm, self).__init__(*args, **kwargs)
+
+    def clean_has_source(self):
+        data = self.cleaned_data
+        if data.get('has_source') == 'yes' and not data.get('source'):
+            raise forms.ValidationError(
+                ugettext(u'You have not uploaded a source file.'))
+        elif data.get('has_source') == 'no' and data.get('source'):
```

Comment:
If a developer clicks "No" after upload sources, we should either clear the upload automatically or give them a way to manually clear. Otherwise you are stuck with selecting "yes" if you have (wrongly) submitted source code.

Reply:
mmm, I didn't account for a developer doing that.  But I guess it's possible.

Reply:
Clear the upload client side?  (what if they toggle between yes and no after selecting the file?)
Or just ignore it server side?

Adding a clear link is possible - but it'd normally be next to the upload selector which would then be hidden once they click 'no' so it's not a great experience.

Reply:
Adding @MeridelW for guidance.

Reply:
I see we don't give users a way to remove the uploaded file in the current experience, either!

I can't speak to whether we should clear the upload on the client vs. server side (I'd think both?), but in terms of the UI, I'd recommend the following:

1. User selects "Yes," uploads the file, and a "Remove" button should appear in place of "No file selected."
2. User selects "Yes," uploads the file, and then selects No. Message is surfaced to user: " "Clear source code file?" buttons "Cancel" or "Yes." If user selects "Yes," source code is cleared and the "No" screen is displayed. If user selects "Cancel," they are returned to "Yes" display.

Would this work? I couldn't find a conditional interaction like this in the current submission flow to use a model...